libs/ml/include/ml/dataloaders/word2vec_loaders/sgns_w2v_dataloader.hpp

```diff
@@ -183,89 +181,147 @@ void W2VLoader<T>::Reset()
-  }
-  for (auto const &sentence : data_)
-  {
-    std::string s;
-    for (auto const &word : sentence)
-    {
-      if (reverse_vocab[word].second >= min)
-      {
-        s += reverse_vocab[word].first + " ";
-      }
+    // remove infrequent words from vocab first
+    vocab_.RemoveInfrequentWord(min);
+
+    // compactify the vocabulary
+    auto old2new = vocab_.Compactify();
+
+    // create reverse vocab
+    auto reverse_vocab = vocab_.GetReverseVocab();
+
+    // create a new data_ for storing text
+    std::vector<std::vector<SizeType>> new_data;
+
+    for(auto sent_it = data_.begin(); sent_it != data_.end(); sent_it++){
+        std::vector<SizeType> new_sent_buffer; // buffer for this sentence
+
+        // reserve every word that is not infrequent
+        for(auto word_it = sent_it->begin(); word_it != sent_it->end(); ++word_it){
+            if(old2new.count(*word_it) > 0) { // if a word is in old2new, append it to the new sentence
+                new_sent_buffer.push_back(*word_it);
+            }
+        }
+
+        // if after we remove infrequent word, the sentence becomes too short, we need to remove the sentence
+        if (new_sent_buffer.size() <= 2*window_size_){
+            for(auto const &word_id : new_sent_buffer) { // reduce the count for all the word in this sentence
+                vocab_.data[reverse_vocab[word_id].first].second -= 1;
+            }
+            // N.B. for practical concerns, we do not further remove infrequent words
+        }else{ // reserve the sentence if the sentence is still long enough
+            new_data.push_back(new_sent_buffer);
+        }
     }
-    new_loader.BuildVocab(s);
-  }
-  data_       = std::move(new_loader.data_);
-  vocab_.data = std::move(new_loader.vocab_.data);
+    data_ = new_data;
+
+    // update vocab
+    vocab_.Update();
+
+    // update dl
+    Update();
 }
 
 /**
  * initialises the unigram table for negative frequency based sampling
  * @tparam T
  */
 template <typename T>
-void W2VLoader<T>::InitUnigramTable()
+void W2VLoader<T>::InitUnigramTable(SizeType size)
 {
   std::vector<SizeType> frequencies(vocab_size());
   for (auto const &kvp : vocab_.data)
   {
     frequencies[kvp.second.first] = kvp.second.second;
   }
-  unigram_table_.Reset(1e8, frequencies);
+  unigram_table_.Reset(frequencies, size);
 }
-
+/**
+ * Buffer for the next word and relevent samples
+ * @tparam T
+ * @param seed
+ */
 template <typename T>
-void W2VLoader<T>::BufferNextSamples()
+void W2VLoader<T>::BufferNextSamples(SizeType seed)
 {
-  // clear the buffer
-  buffer.clear();
+
+    // reset the index to buffer
+    buffer_pos_ = 0;
 
   // the current word should start from position that allows full context window
   if (current_word_ < window_size_)
   {
     current_word_ = window_size_;
   }
 
+  // subsample too frequent word
+  fetch::random::LaggedFibonacciGenerator<> lfg_(seed);
+  while(true){
```

Comment:
break only if IsDone. does this mean the BufferNextSamples buffers an entire epoch?